recipes/CoVoST/AST/train_w2v2_llama.py

```diff
@@ -0,0 +1,550 @@
+        audio_len = wavs.shape[1]
+        abs_len = torch.round(wav_lens * audio_len)
+        audio_attn_mask = length_to_mask(abs_len)
+        wavs = wavs * audio_attn_mask
+
+        # Add waveform augmentation if specified.
+        if (
+            stage == sb.Stage.TRAIN
+            and hasattr(self.hparams, "wav_augment")
+            and self.optimizer_step > self.hparams.augment_warmup
+        ):
+            wavs, wav_lens = self.hparams.wav_augment(wavs, wav_lens)
+
+        wavs = self.hparams.normalize(wavs, wav_lens)
+
+        # Forward Speech Modules
+        feats = self.modules.wav2vec2(wavs, wav_lens)
+        down_feats = self.modules.feat_downsampler(feats)
+        down_feats_proj = self.modules.proj(down_feats)
+
+        # Format input for LLM; [ audio emb ] + [ prompt emb ] + [ translation emb ]
+        # First get relevant lengths
+        audio_len = down_feats_proj.shape[1]
+        text_len = tokens_prompt_translation.shape[1]
+        audio_prompt_len = (audio_len + prompt_len)[0]
+
+        # Then for the full embedding prompt sequence
+        if hasattr(self.modules.lora_llm, "module"):
+            embeddings = (
+                self.modules.lora_llm.module.model.get_input_embeddings()
+            )
+        else:
+            embeddings = self.modules.lora_llm.model.get_input_embeddings()
+
+        inputs_embeds = torch.cat(
+            (down_feats_proj, embeddings(tokens_prompt_translation)), dim=1
+        )
+
+        # Prepare attn_mask for audio and text and combine them. This is not streaming compatible. For HF to work, masked frames should be 0.
+
+        text_abs_len = torch.round(tokens_prompt_translation_len * text_len)
+        abs_len = torch.round(wav_lens * audio_len)
+        audio_attn_mask = length_to_mask(abs_len)
+        text_attn_mask = length_to_mask(text_abs_len)
+        attn_mask = torch.cat([audio_attn_mask, text_attn_mask], dim=-1)
+
+        # LLM forward
+        llm_logits = self.modules.lora_llm(
+            inputs_embeds=inputs_embeds, attention_mask=attn_mask
+        ).logits
+
+        # output layer for seq2seq log-probabilities
+        p_seq = self.hparams.log_softmax(llm_logits)
+
+        if hasattr(self.modules.lora_llm, "module"):
+            gen_func = self.modules.lora_llm.module.model.generate
+        else:
+            gen_func = self.modules.lora_llm.model.generate
+
+        # Running decoding if not training
+        if stage == sb.Stage.TRAIN:
+            hyps = None
+
+        elif stage == sb.Stage.VALID:
+
+            # Define generation config depending on runtime values
+            config = transformers.GenerationConfig(
+                num_beams=self.hparams.valid_beam_size,
+                pad_token_id=tokenizer.pad_token_id,
+                eos_token_id=tokenizer.eos_token_id,
+                max_new_tokens=int(2 * audio_len),
+            )
```

Comment:
can't we define this in the yaml or outside this function?

Reply:
Yeah. I think something should be done on that side. Especially given the fact that this `generate` method has other attributes like `beam_size` etc. It would be nice to have a way to define it elsewere + custom

Reply:
Wait. Why are you doing `max_new_tokens=int(2 * audio_len)` ?

Reply:
AFAIK, we could just plug the LLaMA decoder in our seq2seq and it would/should work with minimal tuning. Usually, we do expect to see in the decoder audio frames, but we could make this optional.

Reply:
I agree on all comments. I don't have time to do it however, and this is non blocking.